presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java

```diff
@@ -230,6 +234,41 @@ private static void generateConstructor(ClassDefinition classDefinition,
                     .putField(sizeField);
         }
 
+        constructor.comment("Set channelArrays field");
+
+        constructor.append(thisVariable.setField(
+                channelArraysField,
+                newInstance(ArrayList.class)));
+
+        if (!channelFields.isEmpty()) {
```

Comment:
Additional overhead might also come from accessing channel in `Block[]`. I have run Yourkit in our benchmark query and majority of overhead comes from `joinFilterFunction` function, so maybe it's worth squeezing additional performance.